bftengine/src/bftengine/KeyExchangeManager.cpp

```diff
@@ -135,15 +136,38 @@ void KeyExchangeManager::loadPublicKeys() {
   notifyRegistry();
 }
 
-void KeyExchangeManager::exchangeTlsKeys(const SeqNum& bft_sn) {
-  auto repId = bftEngine::ReplicaConfig::instance().replicaId;
+void KeyExchangeManager::exchangeTlsKeys(const std::string& type, uint64_t bftSn) {
   auto keys = concord::util::crypto::Crypto::instance().generateECDSAKeyPair(
       concord::util::crypto::KeyFormat::PemFormat, concord::util::crypto::CurveType::secp384r1);
-  auto cert = concord::util::crypto::Crypto::instance().generateSelfSignedCertificate(
-      keys, "node" + std::to_string(repId) + "ser", repId);
-  std::string pk_tmp_file = bftEngine::ReplicaConfig::instance().keyViewFilePath + "/pk.pem." + std::to_string(repId);
-  secretsMgr_->encryptFile(pk_tmp_file, keys.first);
+  std::string root_path =
+      bftEngine::ReplicaConfig::instance().certificatesRootPath + "/" + std::to_string(repID_) + "/" + type;
+
+  std::string cert_path = root_path + "/" + type + ".cert";
+  std::string prev_key_pem = concord::util::crypto::Crypto::instance()
+                                 .RsaHexToPem(std::make_pair(SigManager::instance()->getSelfPrivKey(), ""))
+                                 .first;
+  auto cert = concord::util::crypto::CertificateUtils::generateSelfSignedCert(cert_path, keys.second, prev_key_pem);
+  // Now that we have generated new key pair and certificate, lets do the actual exchange on this replica
+  std::string pk_path = root_path + "/pk.pem";
+  std::fstream nec_f(pk_path);
+  std::fstream ec_f(pk_path + ".enc");
+
+  // 1. exchange private key
+  if (nec_f.good()) {
+    secretsMgr_->encryptFile(pk_path, keys.first);
+    nec_f.close();
+  }
+  if (ec_f.good()) {
+    secretsMgr_->encryptFile(pk_path + ".enc", keys.first);
+    ec_f.close();
+  }
 
+  concord::secretsmanager::SecretsManagerPlain psm_;
+  // 2. exchange the certificate itself
+  psm_.encryptFile(cert_path, cert);
```

Comment:
Same here, we don't encrypt but simply write the cert to the filesystem.

Reply:
the plain secret manager here is used as a file handler (it wraps the read/write operations with error handling and so on). We use it in order to have less code duplications